### PR TITLE
Kubectl retry

### PR DESCRIPTION
**Pull request by pwittrock**
## Pull Request Guidelines
1. Please read our [contributor guidelines](https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md).
2. See our [developer guide](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md).
3. Follow the instructions for [labeling and writing a release note for this PR](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes) in the block below.

``` release-note
Add kubectl flag 'filename-http-retries' to allow retrying failures to get urls passed through '-f'
```

[Analytics]()

### CODE REVIEW

**Comment by pwittrock** [comment]
@ixdy @smarterclayton Is this what you both had in mind for #23747?  (See the second commit)

**Comment by kargakis** [comment]
@kubernetes/kubectl

**Comment by pwittrock** [comment]
fwiw: There are still a few issues with the tests I am working out

File: pkg/kubectl/cmd/util/helpers.go [hunk 1]
```diff
@@ -332,8 +333,28 @@ func AddValidateFlags(cmd *cobra.Command) {
...
+func addHttpRetryFlag(cmd *cobra.Command, options *resource.FilenameParamOptions) {
+	cmd.Flags().IntVar(&options.FilenameHttpRetries, FilenameHttpRetries, 1, "Retry failed http requests for urls passed to '-f' flags this many times.")
+}
+
+func addFlagJsonFilenameFlag(cmd *cobra.Command, options *resource.FilenameParamOptions, usage string) {
+	kubectl.AddJsonFilenameFlag(cmd, &options.Filenames, usage)
+}
+
+func AddFilenameParamFlags(cmd *cobra.Command, options *resource.FilenameParamOptions, usage string) {
+	addFlagJsonFilenameFlag(cmd, options, usage)
```
Comment (janetkuo) [inline pkg/kubectl/cmd/util/helpers.go]: Why not directly call `kubectl.AddJsonFilenameFlag(cmd, &options.Filenames, usage)` here?

Comment (pwittrock) [inline pkg/kubectl/cmd/util/helpers.go]: No real reason.  I can change it to do that.

**Comment by smarterclayton** [comment]
It might be ok to just bake in a 3 times retry on getting any filename urls.  That could be even smaller a change.

**Comment by pwittrock** [comment]
@smarterclayton Yes that is a much smaller change.  Adding a new flag for fileparams is a lot more work than it ought to be.  It would also be simpler to skip the exponential backoff stuff in favor of a constant sleep interval.

I'll send a separate PR for that and close this one for now.  I'd like to get the filename params in a better shape, but it isn't my top priority right now.

File: pkg/kubectl/resource/builder.go [hunk 2]
```diff
@@ -92,14 +92,20 @@ func (b *Builder) Schema(schema validation.Schema) *Builder {
 	return b
 }
 
+type FilenameParamOptions struct {
```
Comment (deads2k) [inline pkg/kubectl/resource/builder.go]: I like structuring the flags with this struct and binding the flags to it.  It makes it easier for composition and consumption.

**Comment by deads2k** [comment]
> It might be ok to just bake in a 3 times retry on getting any filename urls. That could be even smaller a change.

I like the struct and binding the flags to the struct.  Hard-coding a couple retries into that struct at binding time would be fine with me too.

**Comment by smarterclayton** [comment]
Yeah, I'm fine with preserving the struct but just not registering it as a
flag.

**Comment by pwittrock** [comment]
@smarterclayton I changed the assignee to you since you already did a review pass.  Let me know if you would prefer a I change it to someone else.

FWIW PR #25267 is a much simpler alternative that doesn't have the struct updates or exp backoff.  Sounds like we are ok with the approach in this PR, but just don't want to register the flag?  (Leave the struct changes and exponential backoff).  If that is the case I will copy the tests from #25267 and deregister the flag.

**Comment by smarterclayton** [comment]
@deads2k I don't want to expose the flag now (in case it wasn't clear in my response).

**Comment by pwittrock** [comment]
Thinking about this further, if we are not going to expose this as a flag, there is not reason for the flag struct refactoring to be coupled to the retry feature.  WDYT of just going with #25267 and I can change this PR to be a pure flag struct refactoring with no other changes?

**Comment by pwittrock** [comment]
Looks like I need to run the formatters on this...

**Pull request closed without merging**